src/common/simple_cache.hpp

```diff
@@ -22,19 +22,22 @@ template <class K, class V, class C = std::less<K>, class H = std::hash<K> >
 class SimpleLRU {
   Mutex lock;
   size_t max_size;
+  size_t size;
   ceph::unordered_map<K, typename list<pair<K, V> >::iterator, H> contents;
   list<pair<K, V> > lru;
   map<K, V, C> pinned;
 
   void trim_cache() {
-    while (lru.size() > max_size) {
+    while (size > max_size) {
```

Comment:
You're using "size" field as a substitute of "lru.size()" only here, so it'd be safer and easier to make it local for this function, something like
```
size_t size = lru.size();
while (size > max_size) {
   contents.erase(lru.back().first);
   lru.pop_back();
   size--;
}
```
That way you'll save 8 bytes of memory and won't slow down _add() and _clear().

Reply:
that's not true. `size` is updated in `_add()` and `clear()`. but I'd suggest check ` __GLIBCXX__` and `_GLIBCXX_USE_CXX11_ABI` for this workaround. so we keep tracking the `size` only if libstdc++ with old ABI is used. like

```c++
#if defined(__GLIBCXX__) && _GLIBCXX_USE_CXX11_ABI
#define SLOW_LIST_SIZE
#endif
....

#ifdef SLOW_LIST_SIZE
size_t _size;
void inc_size() {
  ++_size;
}
void dec_size() {
  --_size;
}
size_t size() const {
  return _size;
}
#else
void inc_size() {}
void dec_size() {}
size_t size() const {
  return contents.size();
}
#endif
...

void _add(K key, V&& value) {
  ...
  inc_size();
}
```

what do you think?

Reply:
@tchaikov yes, the value itself is *updated* in ```_add()``` and ```_clear()```, but the actual use of it is made in ```trim_cache()``` (also, I think it's missing an update after a call to ```lru.splice()```. My point is - I think that with a such usage pattern, having it count manually will leave extra code that'll need to be removed once we're sure that ```list<>.size()``` is no longer broken.

Reply:
@tchaikov Thanks for your advise.
By the way, do mean "#if defined(__GLIBCXX__) && !_GLIBCXX_USE_CXX11_ABI" instead of "#if defined(__GLIBCXX__) && _GLIBCXX_USE_CXX11_ABI"? I thought it's when _GLIBCXX_USE_CXX11_ABI is 0 that we need a manual count. Right?

Reply:
@branch-predictor oh, thanks, it makes sense!

@xxhdx1985126 ahh, right! updated the code snippet.

Reply:
@branch-predictor Thanks for your advise:-)
I think, since trim_cache() is called for every omap header creation, it could still lead to performance regression when doing massive file/directory creation even if we call list::size() outside the "while" block . Is this right?

Reply:
@xxhdx1985126 Only if you're doing really massive file/directory creation and lru's size is large enough (say, 4096 entries or more). Still, if you do just single size() call per creation, it will be overshadowed by everything else around. Even that can be reduced by calling ```trim_cache()``` randomly (with, say, 33% probability).